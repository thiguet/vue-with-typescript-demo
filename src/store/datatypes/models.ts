interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
