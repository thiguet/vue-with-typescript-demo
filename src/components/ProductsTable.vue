<template>
    <div class="table-container">
        <table id="table" width="100%">
            <tbody>
                <tr
                    class="row"
                    v-for="({ icon, name }, index) in rows"
                    :key="index"
                >
                    <td class="left">
                        <span><img :src="icon"/></span>
                    </td>
                    <td>
                        <span>{{ name }}</span>
                    </td>
                    <td class="right">
                        <div>
                            <Button
                                :id="`edit-${index + 1}`"
                                class="edit"
                                icon="./assets/icons/edit.svg"
                                name="editar"
                                label="Edit"
                                :onclick="evt => onEdit(evt, index)"
                            />
                            <Button
                                :id="`delete-${index + 1}`"
                                class="delete"
                                icon="./assets/icons/delete.svg"
                                name="deletar"
                                label="Delete"
                                :onclick="evt => onDelete(evt, index)"
                            />
                        </div>
                    </td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { ProductsTableComp } from '@/components/models.d';

import Component from 'vue-class-component';

import Button from '@/components/Button.vue';
import { Prop } from 'vue-property-decorator';

interface RowsData {
    icon: string;
    name: string;
}

@Component({
    name: 'ProductsTable',
    components: {
        Button,
    },
})
export default class ProductsTable extends Vue implements ProductsTableComp {
    @Prop({ required: true }) private rows!: Array<RowsData>;

    public onEdit(evt: Event, index: number) {
        return this.$emit('on-edit', index);
    }

    public onDelete(evt: Event, index: number) {
        return this.$emit('on-delete', index);
    }
}
</script>

<style>
#table {
    margin: auto;
    width: 100%;
}

#table > tbody > tr > td > span > img {
    width: 60px;
    filter: brightness(0);
}

.left {
    text-align: 'left';
    margin-left: 1em;
}
.right {
    text-align: 'right';
    margin-right: 1em;
}

.edit {
    justify-content: flex-end !important;
}

.edit > button,
.delete > button {
    background: none;
}

.icon {
    width: 40px;
}
</style>
