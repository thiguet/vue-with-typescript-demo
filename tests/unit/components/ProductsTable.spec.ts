import { mount, shallowMount, Wrapper } from '@vue/test-utils';

import ProductsTable from '@/components/ProductsTable.vue';
import Vue from 'vue';
import { ProductsTableProps } from './models';
import { getFakeProductArray } from '../utils/ProductFactory';

describe('ProductsTable', () => {
    let props: ProductsTableProps;

    const build = () => {
        const options = {
            propsData: props,
        };

        const wrapper: Wrapper<ProductsTable> = shallowMount(
            ProductsTable,
            options,
        );

        const mountedWrapper: Wrapper<ProductsTable> = mount(
            ProductsTable,
            options,
        );

        return {
            wrapper,
            mountedWrapper,
            table: () => mountedWrapper.find('#table'),
            rows: () => mountedWrapper.findAll('.row'),
            editBtn: () =>
                mountedWrapper
                    .findAll('.row')
                    .at(0)
                    .find('#edit-1'),
            deleteBtn: () =>
                mountedWrapper
                    .findAll('.row')
                    .at(0)
                    .find('#delete-1'),
        };
    };

    beforeEach(() => {
        props = {
            rows: getFakeProductArray(),
        };
    });

    it('renders component', () => {
        props = {
            rows: [],
        };
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const { table, rows, editBtn, deleteBtn } = build();

        expect(table().exists()).toBe(true);
        expect(rows().exists()).toBe(true);
        expect(editBtn().exists()).toBe(true);
        expect(deleteBtn().exists()).toBe(true);
    });

    it('dispatch event when we click on the edit btn', async () => {
        const { mountedWrapper, editBtn } = build();

        // await v-for to render our component.

        await Vue.nextTick();

        await editBtn().trigger('click');

        expect(mountedWrapper.emitted('on-edit')).toBeTruthy();
        expect((mountedWrapper.emitted('on-edit') || [])[0]).toStrictEqual([0]);
    });

    it('dispatch event when we click on the delete btn', async () => {
        const { mountedWrapper, deleteBtn } = build();

        // await v-for to render our component.

        await Vue.nextTick();

        await deleteBtn().trigger('click');

        expect(mountedWrapper.emitted('on-delete')).toBeTruthy();
        expect((mountedWrapper.emitted('on-delete') || [])[0]).toStrictEqual([
            0,
        ]);
    });

    it('dispatch event when we click on the edit btn', async () => {
        const { mountedWrapper, editBtn } = build();

        jest.spyOn(mountedWrapper.vm, '$emit');

        // await v-for to render our component.
        await Vue.nextTick();
        await editBtn().trigger('click');

        expect(mountedWrapper.vm.$emit).toBeCalledWith('on-edit', 0);
    });

    it('dispatch event when we click on the delete btn', async () => {
        const { mountedWrapper, deleteBtn } = build();

        jest.spyOn(mountedWrapper.vm, '$emit');

        // await v-for to render our component.
        await Vue.nextTick();
        await deleteBtn().trigger('click');

        expect(mountedWrapper.vm.$emit).toBeCalledWith('on-delete', 0);
    });
});
