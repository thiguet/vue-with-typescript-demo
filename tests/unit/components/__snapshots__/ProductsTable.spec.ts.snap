// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsTable renders component 1`] = `
<div class="table-container">
  <table id="table" width="100%">
    <tbody>
      <tr class="row">
        <td class="left"><span><img></span></td>
        <td><span>test</span></td>
        <td class="right">
          <div>
            <button-stub id="edit-1" name="editar" label="Edit" icon="./assets/icons/edit.svg" onclick="function (evt) { return _vm.onEdit(evt, index); }" class="edit"></button-stub>
            <button-stub id="delete-1" name="deletar" label="Delete" icon="./assets/icons/delete.svg" onclick="function (evt) { return _vm.onDelete(evt, index); }" class="delete"></button-stub>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
